Pushdown Prometheus Predicates #8742

jirislav opened the pull request:
<!-- Thank you for submitting a pull request! Find more information 
at https://trino.io/development/process.html, 
at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md 
and contact us on #core-dev in Slack. -->
<!-- Provide an overview for maintainers and reviewers. -->
## Description

Implements pushdown for Prometheus predicates, as requested in #8742.


<!-- Provide details that help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues



<!-- Mark the appropriate option with an (x). Propose a release note if you can.
More info at https://trino.io/development/process#release-note -->
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
## Prometheus Connector
* Support predicate pushdown for equality, non-equality, LIKE & IN operators. (#8742)
```

jirislav committed:
Push-Down Prometheus Predicates #8742

etolbakov commented on plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/expression/RewriteIn.java:
```diff
@@ -0,0 +1,84 @@
+        List<String> terms = new ArrayList<>(expressions.size());
+        for (ConnectorExpression expression : expressions) {
+            terms.add(extractVarchar(expression));
+        }
```
nit: this section could be rewritten with Java Streams:
```
List<String> terms = expressions.stream()
    .map(this::extractVarchar)
    .collect(Collectors.toList());
 ```
 
 or even (if you consider inlining `expressions`
 ```
 List<String> terms = captures.get(LABEL_VALUES).stream()
    .map(this::extractVarchar)
    .collect(Collectors.toList());
```

jirislav replied:
Great idea. Actually I think we can push it even further and even avoid creating the list:

```java
        String labelName = extractLabelName(captures, LABEL_NAME);
        List<ConnectorExpression> expressions = captures.get(LABEL_VALUES);

        return Optional.of(ImmutableList.of(new LabelFilterExpression(
                labelName,
                expressions.stream()
                        .map(RewriteHelpers::extractVarchar)
                        .collect(Collectors.joining("|", "(", ")")),
                true,
                false)));
```

jirislav commented:
> Thank you for your pull request and welcome to the Trino community. We require contributors to sign our [Contributor License Agreement](https://github.com/trinodb/cla/raw/master/Trino%20Foundation%20Individual%20CLA.pdf), and we don't seem to have you on file. Continue to work with us on the review and improvements in this PR, and submit the signed CLA to [cla@trino.io](mailto:cla@trino.io). Photos, scans, or digitally-signed PDF files are all suitable. Processing may take a few days. The CLA needs to be on file before we merge your changes. For more information, see https://github.com/trinodb/cla

I have the CLA signed and sent like 6 hours ago, not sure how long it usually takes to propagate to the related failing CI job.

jirislav commented:
@ebyhr could you please review & merge? I see you [made recent changes to Prometheus Connector](https://github.com/trinodb/trino/blame/6028e809b8b3993665191766af504e0561607674/plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/PrometheusMetadata.java) as well.

ebyhr commented on plugin/trino-prometheus/src/test/java/io/trino/plugin/prometheus/TestPrometheusIntegration.java:
```diff
@@ -31,8 +33,10 @@
+import static org.junit.jupiter.api.parallel.ExecutionMode.SAME_THREAD;
 
-public class TestPrometheusIntegration
+@Execution(SAME_THREAD)
```
Please leave a code comment why specifying `SAME_THREAD`.

ebyhr commented on plugin/trino-prometheus/src/test/java/io/trino/plugin/prometheus/PrometheusServer.java:
```diff
@@ -14,57 +14,129 @@
     public void close()
     {
-        dockerContainer.close();
+        prometheusContainer.close();
+
+        if (proxyContainer != null) {
+            proxyContainer.close();
+        }
+
+        network.close();
```
You can use Guava's `Closer`.

ebyhr replied:
Try the below code:
```java
private final Closer closer = Closer.create();
...
closer.register(network::close);
```

ebyhr reviewed (commented):
Just skimmed.

chenjian2664 commented on plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/expression/RewriteHelpers.java:
```diff
@@ -0,0 +1,78 @@
+    public static final Predicate<Call> LABEL_MAP_CHECK_CORRECT_TYPE = connectorExpression -> {
+        List<ConnectorExpression> arguments = connectorExpression.getArguments();
+
+        if (!(arguments.getFirst() instanceof Variable)) {
```
Add a check if arguments is empty or not

chenjian2664 commented on plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/expression/RewriteHelpers.java:
```diff
@@ -0,0 +1,78 @@
+        List<ConnectorExpression> arguments = captures.get(captureKey).getArguments();
+        Object labelNameObject = ((Constant) arguments.getLast()).getValue();
+        if (!(labelNameObject instanceof Slice)) {
+            if (labelNameObject == null) {
```
Should we perform this check earlier?

chenjian2664 commented on plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/expression/RewriteHelpers.java:
```diff
@@ -0,0 +1,78 @@
+        // Extract the label name from the call expression, which is actually the key of the labels map.
+        // This method assumes that
+        List<ConnectorExpression> arguments = captures.get(captureKey).getArguments();
+        Object labelNameObject = ((Constant) arguments.getLast()).getValue();
```
add arguments check if it is empty

jirislav committed:
Resolve all compilation warnings for Prometheus Connector

To inspect compilation warnings, see https://github.com/trinodb/trino/actions/runs/15529824872/job/43716197311?pr=25957#step:5:149.

jirislav commented on plugin/trino-prometheus/src/test/java/io/trino/plugin/prometheus/PrometheusServer.java:
```diff
@@ -14,57 +14,129 @@
+    private static final String DEFAULT_NGINX_PROXY_VERSION = "1.28.0-alpine";
+
     private static final int PROMETHEUS_PORT = 9090;
-    private static final String DEFAULT_VERSION = "v2.15.1";
```
You mean `private -> public` or `DEFAULT_VERSION -> DEFAULT_PROMETHEUS_VERSION`?
- The former is needed for this constant to be used in `TestPrometheusIntegration`.
- The latter is needed to avoid ambiguity since we're now using two different docker images, each having different version.

chenjian2664 commented on plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/expression/RewriteIn.java:
```diff
@@ -0,0 +1,79 @@
+                labelName,
+                expressions.stream()
+                        .map(RewriteHelpers::extractVarchar)
+                        .collect(Collectors.joining("|", "(", ")")),
```
nit, we can remove the duplicates before building the value

ramcasa commented:
This is a quite useful feature. Without it querying a decently-sized database is close to impossible. Is there an ETA for merging this?
